### PR TITLE
Add a trigram index on supporter names to improve name searching

### PR DESCRIPTION
**Pull request by clarissalimab**
Using a [trigram index](https://www.postgresql.org/docs/12/pgtrgm.html) on supporter names to improve supporters name search.

Examples of how it behaves now:
- If I wanted to search for a supporter record named "Cacau" and searched for "Caca", it would not be found;
- If I wanted to search for a supporter record named "Cacau" and searched for the full name "Cacau Borges", it would not be found;
- If I wanted to search for a supporter record named "Cacau" and searched for "Caucau", it would not be found.

Using trigram index helps with those cases.

Closes https://github.com/CommitChange/tix/issues/3702.

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -1251,7 +1255,18 @@
...
+                  END
+                $function$
+  SQL
+  create_function :update_name_index_on_supporters, sql_definition: <<-SQL
+      CREATE OR REPLACE FUNCTION public.update_name_index_on_supporters()
+       RETURNS trigger
+       LANGUAGE plpgsql
+      AS $function$
+                  BEGIN
+                    new.name_index = new.name;
```
Comment (clarissalimab) [inline db/schema.rb:1269]: What did you do to avoid this change on your PR?

File: db/schema.rb [hunk 2]
```diff
@@ -1251,7 +1255,18 @@
        LANGUAGE plpgsql
       AS $function$
                   BEGIN
-                    new.phone_index = (regexp_replace(new.phone, '\\D','', 'g'));
+                    new.phone_index = (regexp_replace(new.phone, '\D','', 'g'));
```
Comment (clarissalimab) [inline db/schema.rb]: @wwahammy, what did you do to avoid this change on your PR?

File: lib/query/query_supporters.rb [hunk 3]
```diff
@@ -231,6 +231,8 @@ def self.full_filter_expr(np_id, query)
     if query[:search].present?
       expr = expr.and_where(%Q(
         supporters.fts @@ websearch_to_tsquery('english', $search)
+        OR supporters.name ILIKE $search
```
Comment (clarissalimab) [inline lib/query/query_supporters.rb]: I think that's the reason our name search got worse, we used to compare using `supporters.name ILIKE $search`, but after we added the full text search index, we removed that.

Comment (clarissalimab) [reply lib/query/query_supporters.rb]: We won't need that if the next line is correct, though, I added this back to test.

File: db/schema.rb (hunk 2 above)
Comment (wwahammy) [reply db/schema.rb]: I just had to manually change it back. I need to report the bug to fx

File: db/schema.rb [hunk 4]
```diff
@@ -1267,10 +1269,10 @@
...
   SQL
-  create_trigger :update_supporters_fts, sql_definition: <<-SQL
-      CREATE TRIGGER update_supporters_fts BEFORE INSERT OR UPDATE ON public.supporters FOR EACH ROW EXECUTE FUNCTION update_fts_on_supporters()
-  SQL
   create_trigger :update_supporters_phone_index, sql_definition: <<-SQL
       CREATE TRIGGER update_supporters_phone_index BEFORE INSERT OR UPDATE ON public.supporters FOR EACH ROW EXECUTE FUNCTION update_phone_index_on_supporters()
   SQL
+  create_trigger :update_supporters_fts, sql_definition: <<-SQL
+      CREATE TRIGGER update_supporters_fts BEFORE INSERT OR UPDATE ON public.supporters FOR EACH ROW EXECUTE FUNCTION update_fts_on_supporters()
+  SQL
```
Comment (wwahammy) [inline db/schema.rb:1277]: What happened here?

File: lib/query/query_supporters.rb [hunk 5]
```diff
@@ -231,6 +231,7 @@ def self.full_filter_expr(np_id, query)
     if query[:search].present?
       expr = expr.and_where(%Q(
         supporters.fts @@ websearch_to_tsquery('english', $search)
+        OR supporters.name % $search
```
Comment (wwahammy) [inline lib/query/query_supporters.rb]: Are we sure this leads to useful results? Or do the results become less useful?

**Review by wwahammy — COMMENTED** [review]
I'm curious about how much better the trigram index works in reality. I want to know that we're not making things even worse in some way.

File: db/schema.rb (hunk 4 above)
Comment (clarissalimab) [reply db/schema.rb:1277]: It moved some lines down, I don't know why.

File: lib/query/query_supporters.rb (hunk 5 above)
Comment (clarissalimab) [reply lib/query/query_supporters.rb]: I edited the description with some use cases. I think in general users can distinguish when a result is not what they want, but there's a scenario that we might want to look at:

We have more than 1 Cacau (different people) who donated. One of them has the name "Cacau" in our records, and the other one has "Cacau Borges". I want to find the records from "Cacau Borges", but the results from "Cacau" are going to come up. The emails are going to help them to distinguish the results. A similar behavior already happens if you look for the first name only: you won't find exclusively supporters that match exactly "Cacau", you'll also find "Cacau Borges".

Comment (wwahammy) [reply lib/query/query_supporters.rb]: So, if you search for Cacau Borges, you get "Cacau"? Does it put the records with Cacau Borges first?

Comment (clarissalimab) [reply lib/query/query_supporters.rb]: Not necessarily because right now we're ordering by `last_payment` date. What do you think?

Comment (clarissalimab) [reply lib/query/query_supporters.rb]: There's a way to order by the name similarity.

Comment (wwahammy) [reply lib/query/query_supporters.rb]: I think it would be helpful to have @wendylbolm's viewpoint here.

**Comment by wwahammy** [comment]
What do we need to know to get this merged? @clarissalimab

**Comment by clarissalimab** [comment]
@wwahammy need to ask Wendy's opinion on how it should work to see if it's fine

**Comment by wwahammy** [comment]
@wendylbolm how SHOULD our search work?

**Comment by wwahammy** [comment]
@wendylbolm *waves*

**Comment by wendylbolm** [comment]
I think that having more results than less is good. Also, I don't know if it matters too much the order the results come up in, but it might be less confusing to have them in alphabetical order. 

I think this explanation from Clarissa sounds pretty good: 

We have more than 1 Cacau (different people) who donated. One of them has the name "Cacau" in our records, and the other one has "Cacau Borges". I want to find the records from "Cacau Borges", but the results from "Cacau" are going to come up. The emails are going to help them to distinguish the results. A similar behavior already happens if you look for the first name only: you won't find exclusively supporters that match exactly "Cacau", you'll also find "Cacau Borges".

**Comment by wwahammy** [comment]
Alright, I'm open to merging this if we think it's the best solution.

**Comment by clarissalimab** [comment]
@wwahammy I'm confused, what happened to the commits?

**Comment by clarissalimab** [comment]
Do you happen to know? :laughing:

**Comment by clarissalimab** [comment]
Okay, I think I know what to do here.

**Comment by wwahammy** [comment]
@clarissalimab what's the current state of this PR?

**Comment by clarissalimab** [comment]
@wwahammy it's ready to merge if you approve it.

**Review by wwahammy — APPROVED** [review]
Looks good!

**Pull request merged**